FirebaseAppDistribution/Sources/FIRAppDistribution.m

```diff
@@ -116,181 +123,160 @@ + (instancetype)appDistribution {
-  if (self.window) {
-    self.window.hidden = YES;
-    self.window = nil;
-  }
+- (NSError *)NSErrorForErrorCodeAndMessage:(FIRAppDistributionError)errorCode
+                                   message:(NSString *)message {
+  NSDictionary *userInfo = @{FIRAppDistributionErrorDetailsKey : message};
+  return [NSError errorWithDomain:FIRAppDistributionErrorDomain code:errorCode userInfo:userInfo];
 }
 
-- (void)handleReleasesAPIResponseWithData:data
-                               completion:(FIRAppDistributionUpdateCheckCompletion)completion {
-  NSError *error = nil;
-  NSDictionary *object = [NSJSONSerialization JSONObjectWithData:data options:0 error:&error];
-
-  NSArray *releaseList = [object objectForKey:@"releases"];
-  for (NSDictionary *releaseDict in releaseList) {
-    if (![[releaseDict objectForKey:@"latest"] boolValue]) continue;
-
-    NSString *codeHash = [releaseDict objectForKey:@"codeHash"];
-    FIRAppDistributionMachO *machO =
-        [[FIRAppDistributionMachO alloc] initWithPath:[[NSBundle mainBundle] executablePath]];
-
-    if (![codeHash isEqualToString:machO.codeHash]) {
-      FIRAppDistributionRelease *release =
-          [[FIRAppDistributionRelease alloc] initWithDictionary:releaseDict];
-      dispatch_async(dispatch_get_main_queue(), ^{
-        completion(release, nil);
-      });
+- (NSError *_Nullable)mapFetchReleasesError:(NSError *)error {
+  if ([error domain] == kFIRFADApiErrorDomain) {
+    FIRFADErrorLog(@"Failed to retrieve releases: %ld", (long)[error code]);
+    switch ([error code]) {
+      case FIRFADApiErrorTimeout:
+        return [self NSErrorForErrorCodeAndMessage:FIRAppDistributionErrorNetworkFailure
+                                           message:@"Failed to fetch releases due to timeout."];
+      case FIRFADApiErrorUnauthenticated:
+      case FIRFADApiErrorUnauthorized:
+      case FIRFADApiTokenGenerationFailure:
+      case FIRFADApiInstallationIdentifierError:
+      case FIRFADApiErrorNotFound:
+        return [self NSErrorForErrorCodeAndMessage:FIRAppDistributionErrorAuthenticationFailure
+                                           message:@"Could not authenticate tester"];
+      default:
+        return [self NSErrorForErrorCodeAndMessage:FIRAppDistributionErrorUnknown
+                                           message:@"Failed to fetch releases for unknown reason."];
     }
   }
+
+  FIRFADErrorLog(@"Failed to retrieve releases with unexpected domain %@: %ld", [error domain],
+                 (long)[error code]);
+  return [self NSErrorForErrorCodeAndMessage:FIRAppDistributionErrorUnknown
+                                     message:@"Failed to fetch releases for unknown reason."];
 }
+
+- (void)fetchNewLatestRelease:(FIRAppDistributionUpdateCheckCompletion)completion {
+  [FIRFADApiService
+      fetchReleasesWithCompletion:^(NSArray *_Nullable releases, NSError *_Nullable error) {
+        if (error) {
+          completion(nil, [self mapFetchReleasesError:error]);
```

Comment:
This completion is not wrapped in a dispatch_async, while the one below is. Would recommend always calling the callback on the same thread as it can manifest as super funky bugs down the road.

Reply:
this is done.